pkg/destroy/libvirt/libvirt.go

```diff
@@ -54,102 +49,80 @@ type ClusterUninstaller struct {
 
 // Run is the entrypoint to start the uninstall process.
 func (o *ClusterUninstaller) Run() error {
-	deleteFuncs := map[string]deleteFunc{}
-	populateDeleteFuncs(deleteFuncs)
-	returnChannel := make(chan string)
-
 	conn, err := libvirt.NewConnect(o.LibvirtURI)
 	if err != nil {
 		return errors.Wrap(err, "failed to connect to Libvirt daemon")
 	}
 
-	// launch goroutines
-	for name, function := range deleteFuncs {
-		go deleteRunner(name, function, conn, o.Filter, o.Logger, returnChannel)
-	}
-
-	// wait for them to finish
-	for i := 0; i < len(deleteFuncs); i++ {
-		select {
-		case res := <-returnChannel:
-			o.Logger.Debugf("goroutine %v complete", res)
+	for _, del := range []deleteFunc{
+		deleteDomains,
+		deleteNetwork,
+		deleteVolumes,
+	} {
+		err = del(conn, o.Filter, o.Logger)
+		if err != nil {
+			return err
 		}
 	}
 
 	return nil
 }
 
-func deleteRunner(deleteFuncName string, dFunction deleteFunc, conn *libvirt.Connect, filter filterFunc, logger logrus.FieldLogger, channel chan string) {
-	backoffSettings := wait.Backoff{
-		Duration: time.Second * 10,
-		Factor:   1.3,
-		Steps:    100,
-	}
-
-	err := wait.ExponentialBackoff(backoffSettings, func() (bool, error) {
-		return dFunction(conn, filter, logger)
-	})
-
-	if err != nil {
-		logger.Fatalf("Unrecoverable error/timed out: %v", err)
-		os.Exit(1)
+// deleteDomains calls deleteDomainsSinglePass until it finds no
+// matching domains.  This guards against the machine-API launching
+// additional nodes after the initial list call.  We continue deleting
+// domains until we either hit an error or we have a list call with no
+// matching domains.
+func deleteDomains(conn *libvirt.Connect, filter filterFunc, logger logrus.FieldLogger) error {
+	logger.Debug("Deleting libvirt domains")
+	var err error
+	nothingToDelete := false
+	for !nothingToDelete {
+		nothingToDelete, err = deleteDomainsSinglePass(conn, filter, logger)
+		if err != nil {
+			return err
+		}
 	}
-
-	// record that the goroutine has run to completion
-	channel <- deleteFuncName
-	return
-}
-
-// populateDeleteFuncs is the list of functions that will be launched as goroutines.
-func populateDeleteFuncs(funcs map[string]deleteFunc) {
-	funcs["deleteDomains"] = deleteDomains
-	funcs["deleteVolumes"] = deleteVolumes
-	funcs["deleteNetwork"] = deleteNetwork
+	return nil
 }
 
-func deleteDomains(conn *libvirt.Connect, filter filterFunc, logger logrus.FieldLogger) (bool, error) {
-	logger.Debug("Deleting libvirt domains")
-	defer logger.Debugf("Exiting deleting libvirt domains")
-
+func deleteDomainsSinglePass(conn *libvirt.Connect, filter filterFunc, logger logrus.FieldLogger) (nothingToDelete bool, err error) {
 	domains, err := conn.ListAllDomains(0)
 	if err != nil {
-		logger.Errorf("Error listing domains: %v", err)
-		return false, nil
+		return false, errors.Wrap(err, "list domains")
```

Comment:
https://godoc.org/github.com/pkg/errors#hdr-Adding_context_to_an_error something like `listing domains failed` ?

Reply:
> https://godoc.org/github.com/pkg/errors#hdr-Adding_context_to_an_error something like listing domains failed ?

I think the failure will be clear from the wrapped message and the fact that we're returning an error.  For example:

```console
$ git describe --dirty
v0.3.0-201-g80bb4ed-dirty
$ git diff
diff --git a/vendor/github.com/libvirt/libvirt-go/connect.go b/vendor/github.com/libvirt/libvirt-go/connect.go
index 8cc7cc7..cecea86 100644
--- a/vendor/github.com/libvirt/libvirt-go/connect.go
+++ b/vendor/github.com/libvirt/libvirt-go/connect.go
@@ -1509,6 +1509,7 @@ func (c *Connect) ListAllNetworks(flags ConnectListAllNetworksFlags) ([]Network,
 
 // See also https://libvirt.org/html/libvirt-libvirt-domain.html#virConnectListAllDomains
 func (c *Connect) ListAllDomains(flags ConnectListAllDomainsFlags) ([]Domain, error) {
+       return nil, fmt.Errorf("some error %v", flags)
        var cList *C.virDomainPtr
        var err C.virError
        numDomains := C.virConnectListAllDomainsWrapper(c.ptr, (**C.virDomainPtr)(&cList), C.uint(flags), &err)
$ TAGS=libvirt_destroy hack/build.sh
$ openshift-install --dir=wking --log-level=debug destroy cluster
DEBUG Deleting libvirt domains                     
FATAL Error executing openshift-install: Failed to destroy cluster: list domains: some error 0 
```

I think:

```
FATAL Error executing openshift-install: Failed to destroy cluster: listing domains failed: some error 0
```

would be more redundant.  But I can switch to your wording if you're not convinced.

Reply:
i'm more inclined towards the second option, but could go either way..